Rewrote installation guide to use new reference config (#583)
digital-asset/docs.daml.com: docs/2.8.0/docs/canton/usermanual/persistence.rst

Please note that Canton will create, manage and upgrade the database schema directly. You don't have
to create tables yourselves.

Consult the reference ``config`` directory to get a set of configuration files to set your nodes up.

Postgres
--------

Our reference driver based definition for Postgres configuration is:

.. literalinclude:: /canton/includes/mirrored/community/app/src/pack/config/storage/postgres.conf

You may use this configuration file with environment variables or adapt it accordingly. More detailed setup instructions
and options are available in the `Slick reference guide <https://scala-slick.org/doc/3.3.1/api/index.html#slick.jdbc.JdbcBackend$DatabaseFactoryDef@forConfig(String,Config,Driver,ClassLoader):Database>`_.

> The above configurations are included in the reference ``config/storage`` folder and are sufficient to get going.

Author: fixed

.. _postgres-performance-tuning:

Performance
~~~~~~~~~~~

> Please note that your Postgres database setup requires appropriate tuning to achieve the desired performance. Canton

Author: fixed

> is rather database heavy. This section should give you a starting point for your tuning efforts. You may want to consult

Author: fixed

> the :ref:`troubleshooting section <how_to_measure_db_performance>` on how to analyse whether the database is a limiting factor.

Author: fixed

> This guide can give you a starting point for tuning. Ultimately, every use case is different and the exact resource requirements can not

Reviewer: ```suggestion
This guide can give you a starting point for tuning. Ultimately, every use case is different and the exact resource requirements cannot
```

Author: fixed

be predicted, but have to be measured.

First, ensure that the database you are using is appropriately sized for your use case. The number of cores depends on your
throughput requirements. The rule of thumb is:

- 1 db core per 1 participant core.
- 1 participant core for 30-100 ledger events per second (depends on the complexity of the commands).

The memory requirements depend on your data retention period and the size of the data you are storing. Ideally, you monitor the
database index cache hit/miss ratio. If your instance needs to keep on loading indexes from the disk, performance suffers.
It might make sense to start with 128GB, run a long-running scale & performance test, and `monitor the cache hit/miss ratio <https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STATIO-ALL-INDEXES-VIEW>`__.

Most Canton indexes are contract-id based, which means that the index lookups are randomly distributed. Solid state drives with
high throughput perform much better than spinning disks for this purpose.

> For Postgres, `this online tool <https://pgtune.leopard.in.ua/>`_ is a good starting point for finding reasonable parameters

Reviewer: ```suggestion
For Postgres, `this PGTune online tool <https://pgtune.leopard.in.ua/>`_ is a good starting point for finding reasonable parameters
```

Author: fixed

(use online transaction processing system).

> Beyond the initial configuration, we have made the following observations: Most indexes Canton uses are "hash based".

Reviewer: ```suggestion
Beyond the initial configuration, note that most indexes Canton uses are "hash based".
```

Author: fixed

> Therfore, read and write access to these indexes is uniformly distributed. However, Postgres reads and writes indexes in

Reviewer: ```suggestion
Therefore, read and write access to these indexes is uniformly distributed. However, Postgres reads and writes indexes in
```

Author: fixed

pages of 8kb, while a simple index might only be a couple of writes. Therefore, it is very important to be able to keep the

> indexes in memory and only write updates to the disk from time to time; otherwise, a simple change of 32 bytes requires 8kb IO

Reviewer: ```suggestion
indexes in memory and only write updates to the disk from time to time; otherwise, a simple change of 32 bytes requires 8kb I/O
```

Author: fixed

operations.

> We recommend configuring the ``shared_buffers`` setting to hold 60-70% of the host memory rather than the default

Author: fixed

> suggestion of 25%, as the Postgres caching appears to be more effective than the host based file access caching.

Author: fixed

> We also increase the following variables N times beyond their default: Increase the `checkpoint_timeout` such that

Reviewer: ```suggestion
Also increase the following variables N times beyond their default. Increase the `checkpoint_timeout` so that
```

Author: fixed

the flushing to disk includes several writes and not just one per page, accumulated over time, together with
a higher `max_wal_size` to ensure that the system does not prematurely flush before reaching the checkpoint_timeout

> We recommend that you monitor your system during load testing and tune the parameters accordingly to your use case.

Author: fixed

> The downside of changing the checkpointing parameters is that increase crash recovery takes longer.

Author: fixed
